### ISSUE TITLE
cutlass_fused_moe FP8 incorrect for hidden_size >= 512 (per-tensor dequant path)

### ISSUE THREAD

**Issue by kzjeef** [flashinfer-ai/flashinfer#3068]
## Summary

`cutlass_fused_moe` produces incorrect output when FP8 weights are used with `hidden_size >= 512`. The BF16-only path works correctly for all dimensions. The existing `test_moe_fp8` only tests `hidden_size=128` which masks this issue.

## Test Matrix

All tests use `E=2, top_k=2, I=128, batch_size=1`:

| Input | Weight | H=128 | H=256 | H=512 | H=2048 |
|-------|--------|-------|-------|-------|--------|
| BF16 | BF16 | ✅ | ✅ | ✅ | ✅ (0.37% diff) |
| FP8 | FP8 | ✅ | ✅ | ❌ | ❌ (11% mismatch, max rel 234x) |
| BF16 | FP8 | — | — | ❌ | ❌ crash / all-zero |

## Reproduction

Using FlashInfer v0.6.8rc1 on H200 (SM90):

```python
from tests.moe.test_trtllm_cutlass_fused_moe import test_moe_fp8
import torch

# Existing test — PASS
test_moe_fp8(1, 128, 2, 2, 128, torch.float16, torch.float8_e4m3fn)

# Same test with realistic hidden_size — FAIL
test_moe_fp8(1, 512, 2, 2, 128, torch.float16, torch.float8_e4m3fn)
# → Mismatched elements: 11.1%, greatest abs diff: 1.45, greatest rel diff: 234x

test_moe_fp8(1, 2048, 2, 2, 128, torch.float16, torch.float8_e4m3fn)
# → FAIL
```

The failure is independent of `num_experts` (tested E=2 through E=128, all fail for H>=512).

## Impact

All production FP8 MoE models have `hidden_size >= 2048`:
- Qwen3-30B-A3B: H=2048
- MiniMax M2.5: H=3072
- Qwen3-235B: H=4096
- DeepSeek V3: H=7168

## Notes

- This is in `cutlass_backend` (`cutlass_fused_moe`), not `trtllm_backend` (`trtllm_fp8_block_scale_moe`)
- Related: #2356 was about `trtllm_fp8_block_scale_moe` accuracy and was fixed by #2803 — different code path
- BF16×BF16 path is correct for all dimensions, confirming routing/finalize logic is fine — the bug is specific to the FP8 GEMM or dequant path
- We work around this with auto-fallback to per-expert blockscale GEMM loop

## Environment

- FlashInfer: v0.6.8rc1 (`e843df97`)
- GPU: H200 (SM90)
- CUDA: 12.8

**Comment by RunFMe** [comment]
Hi! I'd like to help with the issue. 
The current test appears to be comparing the FP8 CUTLASS kernel against a PyTorch reference that does all calculation in 16 bit and omits quantizing SwiGLU output to fp8 before gemm2. Until the reference models that boundary, the test can fail even when the kernel is following the intended FP8 MoE computation.

### flashinfer research

CUTLASS Implementation doing the cast: https://github.com/flashinfer-ai/flashinfer/blob/855939764f69b4ad14d43b7440a939947f2f7330/csrc/fused_moe/cutlass_backend/cutlass_fused_moe_kernels.cuh#L2170

To prove that this cast should be included in reference we will take PyTorch implementation of fused moe and check the diff between version with the cast and without it. 
script: https://gist.github.com/RunFMe/3bc5573e6981d490fd892bfd6d67aecf
result:
```
  hidden   seed      max_abs      max_rel   mismatch   plain_norm     fp8_norm  allclose
      64     42     0.015625      1.42892   0.000000      2.12778      2.10344      True
     128     42    0.0355835      31.7652   0.000000      4.90531       4.8489      True
     256     42    0.0546875      1.60352   0.000000      15.2802      15.2966      True
     384     42     0.151367      7.04073   0.002604      37.5299       37.549     False
     512     42     0.168213      14.7377   0.009766      46.4558      46.3395     False
     768     42     0.296875      56.3806   0.014323      96.2091      95.3147     False
    1024     42     0.650879      12.8986   0.074219      189.656      190.552     False
    1536     42     0.615967      2964.27   0.073568      260.017      260.045     False
    2048     42      1.11719      35.2388   0.104980      587.473      588.049     False
    3072     42      1.35938      27.3912   0.113932      710.644      713.444     False
    4096     42      1.84375      163.581   0.114746      1262.73      1258.56     False
```

This suggests that for larger hidden sizes, comparing the CUTLASS fp8 kernel against a PyTorch reference without the intermediate fp8 cast is not a valid correctness check. The reference itself needs to include the FP8 quantization before GEMM2.

So I think the test should first be updated to compare against a fake-quantized PyTorch reference that mirrors the CUTLASS computation. After that, we can separately check whether the CUTLASS implementation itself still has a real numerical issue.

### what other libs do

1. I found the vllm checks its kernels against PyTorch iimplementation which uses fp8 too, so the cast to fp8 before gemm2 is present there. https://github.com/vllm-project/vllm/blob/main/tests/kernels/moe/test_block_fp8.py#L128

**Comment by kzjeef** [comment]
Thanks @RunFMe — you're right. I dug into our DashInfer side:

- Our per-expert FP8 MoE fallback keeps the SwiGLU output in BF16 and runs gemm2 as `BF16 × FP8 → BF16` . No FP8 quantization of the intermediate activation.
- So the `hidden_size>=512` divergence we reported is a **reference-methodology mismatch** between our BF16-mid reference and CUTLASS's FP8-mid-cast design — not a kernel bug. Matches your gist's plain-vs-cast table.

We'll update our side to mirror the CUTLASS / vLLM pattern (FP8 cast of SwiGLU output before gemm2) when we switch back to the `cutlass_fused_moe` path for perf. Our current per-expert fallback is ~strictly more accurate but slower; for production FP8 it's the correct tradeoff to take the speed and eat the same precision loss vLLM tolerates.

Closing this as "reference was wrong, not kernel" from our side. Thanks for the clean diagnosis.
